Switch to newlib nano instead of redlib as C library

Currently, redlib is used as the libc in this project. The disadvantage of this is that redlib is proprietary, and no source code is available. This makes the project hard to build without the mcuxpresso IDE (my end goal is to be able to build the entire project without having to install proprietary tools and to document how to do this). Mcuxpresso also offers two other libc's: newlib and newlib nano. I tried to switch the project to use newlib nano, but found out there's just barely not enough space in SRAM_LOWER_BOTTOM (we need another 72 bytes at least).

Redlib:
```
    PROGRAM_FLASH:        476 KB       494 KB     96.36%
       SRAM_UPPER:       27600 B        64 KB     42.11%
SRAM_LOWER_BOTTOM:       27444 B      27488 B     99.84%
  SRAM_LOWER_AMBE:          0 GB       8744 B      0.00%
   SRAM_LOWER_TOP:          8 KB      29304 B     27.96%
```

Newlib nano:

```
Memory region         Used Size  Region Size  %age Used
/usr/local/mcuxpresso-ide/ide/plugins/com.nxp.mcuxpresso.tools.linux_11.1.0.202001081728/tools/bin/../lib/gcc/arm-none-eabi/8.3.1/../../../../arm-none-eabi/bin/ld: /usr/local/mcuxpresso-ide/ide/plugins/com.nxp.mcuxpresso.tools.linux_11.1.0.202001081728/tools/bin/../lib/gcc/arm-none-eabi/8.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard/libcr_newlib_semihost.a(syscalls.o): in function `_write':
   PROGRAM_FLASH:        476 KB       494 KB     96.36%
      SRAM_UPPER:       27476 B        64 KB     41.93%
SRAM_LOWER_BOTTOM:       27560 B      27488 B    100.26%
 SRAM_LOWER_AMBE:          0 GB       8744 B      0.00%
syscalls.c:(.text._write+0x0): multiple definition of `_write'; ./source/SeggerRTT/Syscalls/SEGGER_RTT_Syscalls_GCC.o:/home/jdv/Projects/OpenGD77-codec2/firmware/Debug_GD-77/../source/SeggerRTT/Syscalls/SEGGER_RTT_Syscalls_GCC.c:110: first defined here
  SRAM_LOWER_TOP:          8 KB      29304 B     27.96%
/usr/local/mcuxpresso-ide/ide/plugins/com.nxp.mcuxpresso.tools.linux_11.1.0.202001081728/tools/bin/../lib/gcc/arm-none-eabi/8.3.1/../../../../arm-none-eabi/bin/ld: firmware.axf section `.data' will not fit in region `SRAM_LOWER_BOTTOM'
/usr/local/mcuxpresso-ide/ide/plugins/com.nxp.mcuxpresso.tools.linux_11.1.0.202001081728/tools/bin/../lib/gcc/arm-none-eabi/8.3.1/../../../../arm-none-eabi/bin/ld: region `SRAM_LOWER_BOTTOM' overflowed by 72 bytes
collect2: error: ld returned 1 exit status
make: *** [makefile:55: firmware.axf] Error 1
```

I pushed my changes on https://github.com/redfast00/OpenGD77/tree/newlib-instead-of-redlib, so you can check it out if you want to.

--- a/firmware/Makefile
+++ b/firmware/Makefile
@@ -1,0 +1,246 @@
+##
+## export PATH=$ARM_TOOLCHAIN_PATH:$PATH
+## mkdir build && cd build && make -f ../Makefile -j8
+##
+
+RADIO := GD77
+
+##
+## Selection of MCU platform and baseband
+##
+ifeq ("$(RADIO)","GD77")
+	DEFINES  := -DPLATFORM_GD77
+else ifeq ("$(RADIO)","GD77s")
+	DEFINES  := -DPLATFORM_GD77S
+else ifeq ("$(RADIO)","DM1801")
+	DEFINES  := -DPLATFORM_DM1801
+endif
+
+##
+## List here your source files (both .s, .c and .cpp)
+##
+SRC := \
+$(wildcard ../source/*.c)                   \
+$(wildcard ../source/dmr_codec/*.c)         \
+../source/dmr_codec/codec_bin.S             \
+$(wildcard ../source/functions/*.c)         \
+$(wildcard ../source/hotspot/*.c)           \
+$(wildcard ../source/hardware/*.c)          \
+$(wildcard ../source/interfaces/*.c)        \
+$(wildcard ../source/io/*.c)                \
+$(wildcard ../source/SeggerRTT/RTT/*.c)     \
+$(wildcard ../source/SeggerRTT/Syscalls/*.c)\
+$(wildcard ../source/usb/*.c)               \
+$(wildcard ../source/user_interface/*.c )
+
+OZERO_SOURCES := \
+../drivers/fsl_dspi.c                                   \
+../source/SeggerRTT/RTT/SEGGER_RTT.c                    \
+../source/SeggerRTT/RTT/SEGGER_RTT_printf.c             \
+../source/SeggerRTT/Syscalls/SEGGER_RTT_Syscalls_GCC.c  \
+../source/dmr_codec/codec.c                             \
+../source/firmware.c                                    \
+../source/functions/sound.c                             \
+../source/hardware/EPL003.c                             \
+../source/hardware/SPI_Flash.c                          \
+../source/hardware/UC1701_transfer.c                    \
+../source/interfaces/i2s.c                              \
+../source/main.c                                        \
+../source/semihost_hardfault.c
+
+##
+## Drivers' source files and include directories
+##
+DRIVERS_INC := \
+-I../drivers                      \
+-I../usb/device/class             \
+-I../usb/device/source            \
+-I../usb/device/include           \
+-I../usb/device/source/ehci       \
+-I../usb/device/source/khci       \
+-I../usb/device/source/lpcip3511  \
+-I../usb/include                  \
+-I../usb/phy 
+
+DRIVERS_SRC := \
+$(wildcard ../drivers/*.c)                     \
+$(wildcard ../usb/device/class/*.c)            \
+$(wildcard ../usb/device/source/*.c)           \
+$(wildcard ../usb/device/source/ehci/*.c)      \
+$(wildcard ../usb/device/source/khci/*.c)      \
+$(wildcard ../usb/device/source/lpcip3511/*.c) \
+$(wildcard ../usb/phy/*.c)                     \
+../osa/usb_osa_freertos.c 
+
+##
+## List here additional static libraries with relative path
+##
+LIBS :=  
+
+##
+## List here additional include directories (in the form -Iinclude_dir)
+##
+INCLUDE_DIRS := \
+-I../source             \
+-I../include            \
+-I../include/hardware   \
+-I../include/codec      \
+-I../include/functions  \
+-I../include/interfaces \
+-I../include/io         \
+-I../include/usb        \
+-I../osa \
+
+##
+## List here additional defines
+##
+DEFINES += \
+-DGITVERSION=\"`git rev-parse --short HEAD`\" \
+-DSDK_DEBUGCONSOLE=0     \
+-DUSE_SEGGER_RTT         \
+-DCR_INTEGER_PRINTF      \
+-DPRINTF_FLOAT_ENABLE=0  \
+-DFSL_RTOS_FREE_RTOS     \
+-DSDK_OS_FREE_RTOS       \
+-DCPU_MK22FN512VLL12_cm4 \
+-D__MCUXPRESSO           \
+-D__USE_CMSIS            \
+-DDEBUG                  \
+
+
+##
+## Debug defines
+##
+
+DEBUG := 1
+ifeq ($(DEBUG), 1)
+    DEFINES += -DDEBUG -g3
+else
+    DEFINES += -DNDEBUG
+endif
+
+
+##
+## Define used to select target processor
+##
+TARGET := -DCPU_MK22FN512VLL12
+
+
+##
+## Optimization level
+##
+OPTLEVEL := -Os
+#OPTLEVEL:= -O2
+#OPTLEVEL:= -O3
+#OPTLEVEL:= -Os
+
+##
+## Device-specific source files and include directories, e.g. startup code
+##
+DEVICE_INC := -I../CMSIS -I../board -I../device
+DEVICE_SRC :=  \
+$(wildcard ../board/*.c)      \
+../device/system_MK22F51212.c \
+$(wildcard ../startup/*.c) 
+
+##
+## Operating system's source files and include directories
+##
+OS_INC := \
+-I../amazon-freertos/include           \
+-I../amazon-freertos/FreeRTOS/portable 
+
+OS_SRC := \
+$(wildcard ../amazon-freertos/FreeRTOS/*.c)                 \
+../amazon-freertos/FreeRTOS/portable/fsl_tickless_systick.c \
+../amazon-freertos/FreeRTOS/portable/heap_4.c               \
+../amazon-freertos/FreeRTOS/portable/port.c
+
+ALL_INC := $(OS_INC) $(DEVICE_INC) $(DRIVERS_INC) $(INCLUDE_DIRS)
+ALL_SRC := $(SRC) $(OS_SRC) $(DEVICE_SRC) $(DRIVERS_SRC)
+CONFIGS := $(TARGET)  \
+			-fno-common -g3 -Wall -ffunction-sections   \
+			-fdata-sections -ffreestanding -fno-builtin \
+			-fmerge-constants
+
+REDLIB := 0
+ifeq ($(REDLIB), 1)
+    DEFINES += -D__REDLIB__
+    LINKERSCRIPT = ../linkerscripts/firmware.ld
+    CONFIGS += -specs=redlib.specs
+else
+    DEFINES += -D__NEWLIB_H__
+    LINKERSCRIPT = ../linkerscripts/firmware_newlib.ld
+endif
+
+
+ifeq ("$(VERBOSE)","1")
+Q := 
+ECHO := @true
+else
+Q := @
+ECHO := @echo
+endif
+
+
+BUILD_DIR ?= ./bin
+
+## Replaces both "foo.cpp"-->"foo.o" and "foo.c"-->"foo.o"
+OBJ := $(addprefix $(BUILD_DIR)/,  $(addsuffix .o, $(basename $(ALL_SRC))))
+OZERO_OBJECTS := $(addprefix $(BUILD_DIR)/,  $(addsuffix .o, $(basename $(OZERO_SOURCES))))
+
+CFLAGS   := $(ALL_INC) -mcpu=cortex-m4 -mfpu=fpv4-sp-d16 -mfloat-abi=hard -mthumb $(CONFIGS) \
+            $(DEFINES) $(EXCEPT) -c
+
+AFLAGS   := -mcpu=cortex-m4 -mfpu=fpv4-sp-d16 -mfloat-abi=hard -mthumb
+LFLAGS   := -nostdlib -Xlinker -Map="firmware.map" -Xlinker --gc-sections -Xlinker \
+			-print-memory-usage -Xlinker --sort-section=alignment -mcpu=cortex-m4  \
+			-mfpu=fpv4-sp-d16 -mfloat-abi=hard -mthumb -T $(LINKERSCRIPT)
+
+DFLAGS   := -MMD -MP
+
+LINK_LIBS := $(LIBS)
+
+CC  := arm-none-eabi-gcc
+AS  := arm-none-eabi-as
+CP  := arm-none-eabi-objcopy
+SZ  := arm-none-eabi-size
+
+all: OpenGD77.sgl
+
+OpenGD77.sgl: firmware.axf
+	$(ECHO) 'Performing post-build steps'
+	$(Q)$(SZ) $(BUILD_DIR)/firmware.axf
+	$(Q)$(CP) -v -O binary $(BUILD_DIR)/firmware.axf $(BUILD_DIR)/firmware.bin
+	$(Q) ../tools/bin2sgl $(BUILD_DIR)/firmware.bin
+	$(Q) mv $(BUILD_DIR)/firmware.sgl $(BUILD_DIR)/OpenGD77.sgl
+#../tools/GD77_FirmwareLoader "firmware.bin" GUI
+# checksum -p MK22FN512xxx12 -d "firmware.bin"
+
+firmware.axf: $(OBJ) #all-recursive
+	$(ECHO) "[LD ] firmware.axf"
+	$(Q)$(CC) $(LFLAGS) -o $(BUILD_DIR)/firmware.axf $(OBJ) $(LINK_LIBS)
+
+$(OZERO_OBJECTS):
+	$(Q)$(MKDIR_P) $(dir $@)
+	$(ECHO) "[CC0] $(@:bin/%.o=%.c)"
+	$(Q)$(CC)  $(DFLAGS) $(CFLAGS) -O0 $(@:bin/%.o=%.c) -o $@
+
+$(BUILD_DIR)/%.o: %.s
+	$(Q)$(MKDIR_P) $(dir $@)
+	$(ECHO) "[AS ] $<"
+	$(Q)$(AS)  $(AFLAGS) $< -o $@
+
+$(BUILD_DIR)/%.o : %.c
+	$(Q)$(MKDIR_P) $(dir $@)
+	$(ECHO) "[CC ] $<"
+	$(Q)$(CC)  $(DFLAGS) $(CFLAGS) $(OPTLEVEL) $< -o $@
+
+.PHONY: clean
+
+clean:
+	$(Q)$(RM) -rf $(BUILD_DIR) firmware.map && find . -type f -name "*.o" -exec rm {} + -o -name "*.d" -exec rm {} +
+
+-include $(DEPS)
+
+MKDIR_P ?= mkdir -p

--- a/firmware/linkerscripts/firmware_library_newlib.ld
+++ b/firmware/linkerscripts/firmware_library_newlib.ld
@@ -1,0 +1,10 @@
+/*
+ * Generated Linker script file for MK22FN512xxx12
+ */
+
+GROUP (
+  "libgcc.a"
+  "libnosys.a"
+  "libc_nano.a"
+  "libm.a"
+)

--- a/firmware/linkerscripts/firmware_newlib.ld
+++ b/firmware/linkerscripts/firmware_newlib.ld
@@ -1,0 +1,210 @@
+/*
+ * Generated Linker script file for MK22FN512xxx12
+ */
+
+INCLUDE "../linkerscripts/firmware_library_newlib.ld"
+INCLUDE "../linkerscripts/firmware_memory.ld"
+
+ENTRY(ResetISR)
+
+SECTIONS
+{
+    /* MAIN TEXT SECTION */
+    .text_start : ALIGN(4)
+    {
+        FILL(0xff)
+        __vectors_start__ = ABSOLUTE(.) ;
+        KEEP(*(.isr_vector))
+        /* Global Section Table */
+        . = ALIGN(4) ;
+        __section_table_start = .;
+        __data_section_table = .;
+        LONG(LOADADDR(.data));
+        LONG(    ADDR(.data));
+        LONG(  SIZEOF(.data));
+        LONG(LOADADDR(.data_RAM4));
+        LONG(    ADDR(.data_RAM4));
+        LONG(  SIZEOF(.data_RAM4));
+        __data_section_table_end = .;
+        __bss_section_table = .;
+        LONG(    ADDR(.bss));
+        LONG(  SIZEOF(.bss));
+        LONG(    ADDR(.bss_RAM2));
+        LONG(  SIZEOF(.bss_RAM2));
+        LONG(    ADDR(.bss_RAM4));
+        LONG(  SIZEOF(.bss_RAM4));
+        __bss_section_table_end = .;
+        __section_table_end = . ;
+        /* End of Global Section Table */
+
+        *(.after_vectors*)
+    } > PROGRAM_FLASH
+
+    .codec_bin_section_1 : ALIGN(4)
+    {
+       . = ABSOLUTE(0x4400) ;
+       *(.codec_bin_section_1)
+        KEEP(*(.codec_bin_section_1))
+    } > PROGRAM_FLASH
+
+    .text_main : ALIGN(4)
+    {
+       *(.text*)
+       *(.rodata .rodata.* .constdata .constdata.*)
+       . = ALIGN(4);
+       _etext = . ;
+    } > PROGRAM_FLASH
+    
+    /* DATA section for SRAM_LOWER */
+    .data_RAM4 : ALIGN(4)
+    {
+        FILL(0xff)
+        PROVIDE(__start_data_RAM2 = .) ;
+        *(.ramfunc.$RAM4)
+        *(.ramfunc.$SRAM_LOWER_TOP)
+        *(.data.$RAM4*)
+        *(.data.$SRAM_LOWER_TOP*)
+        . = ALIGN(4) ;
+        PROVIDE(__end_data_RAM2 = .) ;
+     } > SRAM_LOWER_TOP AT>PROGRAM_FLASH
+     
+    /* DATA section for SRAM_LOWER */
+    .data_AMBE : ALIGN(4)
+    {
+        FILL(0xff)
+        PROVIDE(__start_data_RAM3 = .) ;
+        *(.ramfunc.$AMBE)
+        *(.ramfunc.$SRAM_LOWER_AMBE)
+        *(.data.$AMBE*)
+        *(.data.$SRAM_LOWER_AMBE*)
+        . = ALIGN(4) ;
+        PROVIDE(__end_data_RAM3 = .) ;
+     } > SRAM_LOWER_AMBE
+     
+         /* Main DATA section (SRAM_UPPER) */
+    .data : ALIGN(4)
+    {
+       FILL(0xff)
+       _data = . ;
+       *(vtable)
+       *(.ramfunc*)
+       *(.data*)
+       . = ALIGN(4) ;
+       _edata = . ;
+    } > SRAM_UPPER AT>PROGRAM_FLASH
+     
+    /* MAIN DATA SECTION */
+    .uninit_RESERVED : ALIGN(4)
+    {
+        KEEP(*(.bss.$RESERVED*))
+        . = ALIGN(4) ;
+        _end_uninit_RESERVED = .;
+    } > SRAM_UPPER
+
+
+    .codec_bin_section_2 : ALIGN(4)
+    {
+        . = ABSOLUTE(0x50000) ;
+       *(.codec_bin_section_2)
+        KEEP(*(.codec_bin_section_2))
+    } > PROGRAM_FLASH
+
+    /* BSS section for SRAM_LOWER */
+    .bss_RAM2 : ALIGN(4)
+    {
+       PROVIDE(__start_bss_RAM2 = .) ;
+       *(.bss.$RAM2*)
+       *(.bss.$SRAM_LOWER_BOTTOM*)
+       *(COMMON)
+       . = ALIGN (. != 0 ? 4 : 1) ; /* avoid empty segment */
+       PROVIDE(__end_bss_RAM2 = .) ;
+    } > SRAM_LOWER_BOTTOM 
+
+    /* BSS section for SRAM_LOWER */
+    .bss_RAM4 : ALIGN(4)
+    {
+       PROVIDE(__start_bss_RAM4 = .) ;
+       *(.bss.$RAM4*)
+       *(.bss.$SRAM_LOWER_TOP*)
+       . = ALIGN (. != 0 ? 4 : 1) ; /* avoid empty segment */
+       PROVIDE(__end_bss_RAM4 = .) ;
+    } > SRAM_LOWER_TOP 
+
+    /* MAIN BSS SECTION */
+    .bss : ALIGN(4)
+    {
+        _bss = .;
+        *(.bss*)
+        *(COMMON)
+        . = ALIGN(4) ;
+        _ebss = .;
+        PROVIDE(end = .);
+    } > SRAM_UPPER
+
+    /* NOINIT section for SRAM_LOWER_BOTTOM */
+    .noinit_RAM2 (NOLOAD) : ALIGN(4)
+    {
+       *(.noinit.$RAM2*)
+       *(.noinit.$SRAM_LOWER_BOTTOM*)
+       . = ALIGN(4) ;
+    } > SRAM_LOWER_BOTTOM 
+    
+    /* NOINIT section for SRAM_LOWER_AMBE */
+    .noinit_AMBE (NOLOAD) : ALIGN(4)
+    {
+       *(.noinit.$AMBE*)
+       *(.noinit.$SRAM_LOWER_AMBE*)
+       . = ALIGN(4) ;
+    } > SRAM_LOWER_AMBE 
+
+    /* NOINIT section for SRAM_LOWER_TOP */
+    .noinit_RAM4 (NOLOAD) : ALIGN(4)
+    {
+       *(.noinit.$RAM4*)
+       *(.noinit.$SRAM_LOWER_TOP*)
+       . = ALIGN(4) ;
+    } > SRAM_LOWER_TOP 
+
+    /* DEFAULT NOINIT SECTION */
+    .noinit (NOLOAD): ALIGN(4)
+    {
+        _noinit = .;
+        *(.noinit*) 
+         . = ALIGN(4) ;
+        _end_noinit = .;
+    } > SRAM_UPPER
+
+    /* Reserve and place Heap within memory map */
+    _HeapSize = 0x1000;
+    .heap :  ALIGN(4)
+    {
+        _pvHeapStart = .;
+        . += _HeapSize;
+        . = ALIGN(4);
+        _pvHeapLimit = .;
+    } > SRAM_LOWER_TOP
+
+     _StackSize = 0x1000;
+     /* Reserve space in memory for Stack */
+    .heap2stackfill  :
+    {
+        . += _StackSize;
+    } > SRAM_LOWER_TOP
+    
+    /* Locate actual Stack in memory map */
+    .stack ORIGIN(SRAM_LOWER_TOP) + LENGTH(SRAM_LOWER_TOP) - _StackSize - 0:  ALIGN(4)
+    {
+        _vStackBase = .;
+        . = ALIGN(4);
+        _vStackTop = . + _StackSize;
+    } > SRAM_LOWER_TOP
+
+    /* Provide basic symbols giving location and size of main text
+     * block, including initial values of RW data sections. Note that
+     * these will need extending to give a complete picture with
+     * complex images (e.g multiple Flash banks).
+     */
+    _image_start = LOADADDR(.text_start);
+    _image_end = LOADADDR(.codec_bin_section_2) + SIZEOF(.codec_bin_section_2);
+    _image_size = _image_end - _image_start;
+}

